wip: new(pkg,test): added an `integration_test` target to makefile.

FedeDP opened the pull request:
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area cmd

> /area pkg

> /area docs

/area tests


**What this PR does / why we need it**:

Added integration tests to CI: basically, we store a bunch of config files and try to build drivers for them.
Note: i uploaded various needed packages in test/ folder to make sure that they're always available.

Moreover, dropped requirement of 2 urls for amazonlinux (i could build various kernels without it; **more testing needed**!).
Finally, added support for relative urls in `getResolvingURLs` function (needed as kernel-crawler does not resolve relative urls, eg: for amazonlinux2).

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

NOTE: i added a bunch of devel packages (deb,rpm) to make sure that tests still work even if kernel-devel upstream link are taken down.

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
new: added integration tests to CI jobs for x86_64 and arm64
```

FedeDP commented on pkg/driverbuilder/builder/amazonlinux.go:
```diff
@@ -119,23 +119,20 @@ func script(c Config, targetType Type) (string, error) {
-    if err != nil {
-        return "", err
-    }
-    if len(urls) < 2 {
```
Dropped this. Do we really need kernel package?
In my tests, it was not needed.

dwindsor replied:
Why has this check been removed?

FedeDP commented:
Closing and reopening to trigger poiana.
/close

poiana commented:
@FedeDP: Closed this PR.

<details>

In response to [this](https://github.com/falcosecurity/driverkit/pull/160#issuecomment-1157715222):

>Closing and reopening to trigger poiana.
>/close


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

FedeDP commented:
/reopen

poiana commented:
@FedeDP: Reopened this PR.

<details>

In response to [this](https://github.com/falcosecurity/driverkit/pull/160#issuecomment-1157715348):

>/reopen


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

FedeDP commented on pkg/driverbuilder/builder/amazonlinux.go:
```diff
@@ -291,8 +288,7 @@ func fetchAmazonLinuxPackagesURLs(kv kernelrelease.KernelRelease, targetType Typ
 		}
 
 		// Found, do not continue
-		// todo > verify amazonlinux always needs 2 packages (kernel and kernel-devel) too
```
Dropped this; in my tests, it was not needed.

maxgio92 commented:
@FedeDP this is awesome and very useful. The AWS kernel-devel URL are supposed to fail in order to test against missing packages?
Because they remove them from time to time (don't know the lifecycle).

FedeDP commented:
@maxgio92 
I added all the required devel packages to the repo:
> NOTE: i added a bunch of devel packages (deb,rpm) to make sure that tests still work even if kernel-devel upstream link are taken down.

This way, we can assure that the packages are always present!

FedeDP commented on pkg/driverbuilder/docker.go:
```diff
@@ -98,7 +98,8 @@ func mustCheckArchUseQemu(ctx context.Context, b *builder.Build, cli *client.Cli
 	err = cli.ContainerStop(ctx, qemuImage.ID, nil)
-	if err != nil {
+	// Do not leave here if container is already stopped
+	if err != nil && !client.IsErrNotFound(err) {
```
This is a possible fix for qemu cross builds.

dwindsor replied:
Hmm... is this required for the new `integration_test` target to work?

FedeDP commented:
/cc @dwindsor :)

FedeDP commented on .circleci/config.yml:
```diff
@@ -14,6 +16,9 @@ jobs:
           command: make test
+      - run:
+          name: Integration tests # test cross build too!
+          command: make -e TARGET_TEST_ARCH=* integration_test
```
On x86_64 we want to test cross build through qemu too!

dwindsor commented on pkg/driverbuilder/builder/builders.go:
```diff
@@ -50,9 +52,26 @@ func moduleDownloadURL(c Config) string {
 	return fmt.Sprintf("%s/%s.tar.gz", c.DownloadBaseURL, c.DriverVersion)
 }
 
+func resolveURLReference(u string) string {
```
Could we make this a separate issue/PR please?

FedeDP commented:
Note: this requires:
* https://github.com/falcosecurity/driverkit/pull/166
* https://github.com/falcosecurity/driverkit/pull/167
* https://github.com/falcosecurity/driverkit/pull/168

to be merged before.

poiana commented:
LGTM label has been added.  <details>Git tree hash: e560a898a56eca617da216f438b5b847590fab6c</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/160#pullrequestreview-1017116758" title="Approved">dwindsor</a>*, *<a href="https://github.com/falcosecurity/driverkit/pull/160#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)~~ [dwindsor]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

dwindsor commented:
/lgtm

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/driverkit/pull/160#issuecomment-1164482427" title="LGTM">dwindsor</a>*, *<a href="https://github.com/falcosecurity/driverkit/pull/160#" title="Author self-approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fdriverkit).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/driverkit/blob/master/OWNERS)~~ [dwindsor]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
